Support advertising Pod CIDRs via BGP

christarazi opened the pull request:
This pull request implements the ability to advertise a node's Pod CIDR configuration via BGP.

In doing so modifications are made to our k8s event handling infrastructure. The most notable change being the support for "Notification Chains".  This pattern is described further in this issue: https://github.com/cilium/cilium/issues/15471 

A new flag is made available to the Agent for configuring Pod CIDR BGP announcements: 

```
 --bgp-announce-pod-cidr                                Announces the node's pod CIDR via BGP
```

Fixes: https://github.com/cilium/cilium/issues/15015

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -400,8 +400,11 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	d.redirectPolicyManager = redirectpolicy.NewRedirectPolicyManager(d.svc)
-	if option.Config.BGPAnnounceLBIP {
-		d.bgpSpeaker = speaker.New()
+	if option.Config.BGPAnnounceLBIP || option.Config.BGPAnnouncePodCIDR {
```
@christarazi 

If we keep the BGP configuratoin flat like this, this or statement is going to continue to grow. 

This maybe out of scope for this PR, but should bgp config in daemon become structured? 

```
type BGP struct {
  AnnounceLBIP  bool
  AnnouncePodCIDR bool
}
```

And this struct gets attached to the daemon config object. Thoughts?

christarazi replied:
In general yes I think that's a good suggestion. However, I don't feel too strongly about it because we know that we only have 3 use cases for BGP--LB IP, PodCIDR, and Static Egress IP. So in the end, the refactoring doesn't have that much of an impact, IMO.

christarazi replied:
I guess I can see that you might have taken inspiration from `speaker.Opts`. If you do go ahead with the refactoring, then I'd be fine with it. At least the code structure gets cleaned up a bit :+1:

ldelossa replied:
noted: low hanging fruit so I will handle the rest of the PR first and keep this refactor in mind.

ldelossa commented on pkg/endpoint/endpoint.go:
```diff
@@ -1759,6 +1759,28 @@ func (e *Endpoint) UpdateLabels(ctx context.Context, identityLabels, infoLabels
 
+// UpdateLabelsFrom is a convenience function to update an endpoint's identity
+// labels from any source.
+func (e *Endpoint) UpdateLabelsFrom(oldLbls, newLbls map[string]string, source string) error {
```
nice - this is where I think this logically belongs anyway.

ldelossa commented on pkg/k8s/watchers/subscriber/node.go:
```diff
@@ -0,0 +1,86 @@
+)
+
+// NodeHandler is implemented by event handlers responding to K8s Node events.
+type NodeHandler interface {
```
@christarazi 

While this is a bit subjective, my initial read-thru was confused by this terminology of "NodeHandler". In context with the package name we get "subscriber.NodeHandler". Handler to me makes me think HTTP handler not a "subscriber". 

I'm not too hell-bent on this but if you are willing to consider alternative names here are a few that would have helped me grok this easier. 

"subscriber.NodeEvents"
"subscriber.Node"
"subscriber.NodeSubscriber" 

With those names it becomes apparent that "okay, if I want to write a subscriber which listens for Node events, I want to implement a {one of the above}". 

again, subjective, but let me know if you're open to thinking thru the naming.

christarazi replied:
Yes, that's a good suggestion. I would definitely be open to improving the naming. In my head, I was kinda interchanging "handler" and "subscriber", so we should just go with the latter everywhere.

ldelossa replied:
Which of the above naming suggestion do you think is best? Feel free to suggest others as well.

christarazi replied:
I personally like "subscriber.Node" the best because "subscriber.NodeSubscriber" stutters (redundant "subscriber") and the first one I don't see the value of saying "Events". The middle one is sufficient.

ldelossa replied:
Agreed

ldelossa commented on pkg/k8s/watchers/subscriber/node.go:
```diff
@@ -0,0 +1,86 @@
+
+// NodeList holds the NodeHandler subscribers that are notified when reacting
+// to K8s Node resource / object changes in the K8s watchers.
+type NodeList struct {
```
@christarazi one other naming suggestion, to keep semantics related to pub/sub how do you feel about calling this a "chain" instead of "list". 

Im loosely borrowing the terms used in notification chain semantics here: https://0xax.gitbooks.io/linux-insides/content/Concepts/linux-cpu-4.html as a model. 

Youd then wind up with the nice semantics of "k8sWatcher.NodeChain.Register"

christarazi replied:
I'm fine with this renaming :+1:

ldelossa replied:
@christarazi one other note here:

The "NodeChain" could itself implement "subscriber.Node" interface - this would be similar to how golang's ServeMux implements http.Handler. The serve mux itself is a handler, which will register and call other handlers. see: https://golang.org/src/net/http/server.go?s=76212:76272#L2429

In the same vein the "NodeChain" itself can be a subscriber.Node - allowing the method set to stay the same across k8sWatcher usage and implementations.

christarazi reviewed (commented):
A few suggestions:

- Commit "bgp: refactor method set to return errs" seems like it can be squashed with commit "k8s/watchers/subscriber: Allow errors to be returned" 
- The two unit tests commits can be squashed together
- Commit "docs: update documentation" can likely be squashed with commit "daemon, option: Add pod CIDR via BGP option "

Any new file with the old copyright format should be switched to the more compact form. 

By the way, my commits weren't intended to be final; feel free to amend them. This is why I suggest squashing. I don't see the need for this PR to be 18 commits :wink:. Also, feel free to add yourself as co-author by amending the commit msgs with `Co-authored-by: name <email>`.

ldelossa committed:
vendor: metallb bump

this patch bumps our metallb fork to the necessary version to support
podcidr announcements.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>
Co-authored-by: Chris Tarazi <chris@isovalent.com>

ldelossa committed:
bgp,manager: refactor bgp manager and test

this patch reworks the bgp manager to support unit testing, plumb
context correctly, and disambiguate its method set.

files which depend on the manager are updated to reflect these changes.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>
Co-authored-by: Chris Tarazi <chris@isovalent.com>

ldelossa committed:
bgp,speaker: implement podcidr annoucements

this patch implements the ability for our bgp speaker to announce
podCidrs.

a new notifications chain pattern is put into place to support this use
case.

k8s infrastructure is updated to reflect this new notification chain
pattern along with supporting podcidr announcement.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>
Co-authored-by: Chris Tarazi <chris@isovalent.com>

ldelossa committed:
docs: update docs for podcidr announcements

this patch updates the cli and helm chart documentation to support
configuring Cilium to announce podcidrs via bgp.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

christarazi commented on daemon/cmd/daemon_main.go:
```diff
@@ -1454,6 +1464,7 @@ func initEnv(cmd *cobra.Command) {
 			log.Warn("netfilter-compatible-mode is enabled only on 5.4 and higher versions of kernel.")
 		}
 	}
+
```
Extra space?

ldelossa replied:
Don't think so, this just separates the if block from the final bracket of the highest scope.

christarazi commented on pkg/bgp/speaker/metallb.go:
```diff
@@ -0,0 +1,107 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
New copyright format?

christarazi reviewed (commented):
Two minor comments. For copyright, there seems to be many new files with the old format.

joamaki commented on pkg/bgp/manager/metallb.go:
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021 Authors of Cilium
+
+// Package speaker abstracts the BGP speaker controller from MetalLB. This
```
Comment talks about "speaker" package. Check that the comment is up-to-date and reflects the package function correctly.

joamaki commented on pkg/bgp/speaker/events.go:
```diff
@@ -49,18 +69,57 @@ func (s *Speaker) run() {
+		}
+	}
+	if s.announcePodCIDR {
+		if err := s.announcePodCIDRs(*k.podCIDRs); err != nil {
```
Should we skip this if `failed = true` already?

christarazi replied:
No because announcing Pod CIDRs and LB IPs are different cases that shouldn't be tied together. They're just being handled in the same function.

joamaki replied:
:+1:

joamaki commented on pkg/bgp/speaker/events.go:
```diff
@@ -49,18 +69,57 @@ func (s *Speaker) run() {
+
+	if s.announceLBIP {
+		if r := s.speaker.SetNodeLabels(*k.labels); r != types.SyncStateSuccess {
+			failed = true
```
Should this be logged as in the case below?

ldelossa replied:
I would think so, fixing this :+1:

joamaki commented on pkg/bgp/speaker/events.go:
```diff
@@ -49,18 +69,57 @@ func (s *Speaker) run() {
+		}
+	}
+
+	if failed {
```
unimportant nit: could simplify this if `ret = types.SyncStateSuccess` at beginning and doing just `return ret` here

ldelossa replied:
@christarazi, @joamaki comment makes me think here. 

This function `handleNodeEvent` has two side effects:
- set node labels 
- announce pod cidr

Because this forms a permutation of results, we wind up in the situation where we may return "SyncStateFailed" even when only one of those side effects have occurred. This feels wrong to me. 

I think we may want to realize the `handleNodeEvent` as a atomic operation (loosely speaking) where if any of the configured side effects fails, we bail immediately. What do you think Chris and Jussi?

ldelossa replied:
looking at this a bit more, I'm noticing it is this way because we reuse MetalLB's type `SyncState` as MetalLB doesn't return error codes proper, but seems to summarize things with that enum. That's a little unfortunate because it lacks detail for us. 

That being said, its probably fine for now, the only real rub with this is that if say, announcing the pod cidr fails in a loop, but setting the node labels is successful, we will continually set node labels everytime we retry the pod cidr announcement.

christarazi replied:
Yeah it's not as clean as it should be, but the risk for problems by setting the node labels in a loop while we retry to announce the podcidr is very low.

joamaki replied:
Yeah, I'm fine with this.

joamaki commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -164,15 +198,60 @@ func (s *Speaker) OnUpdateEndpointSliceV1Beta1(eps *slim_discover_v1beta1.Endpoi
+	s.queue.Add(nodeEvent{
+		labels:   nodeLabels(node.Labels),
+		podCIDRs: podCIDRs(node),
+		withDraw: t,
```
why not just `true` here?

ldelossa replied:
lol, I am not actually sure. Will update.

joamaki commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -273,3 +352,25 @@ func convertEndpointSliceV1Beta1(in *slim_discover_v1beta1.EndpointSlice) *metal
+		return nil
+	}
+	podCIDRs := make([]string, 0, len(node.Spec.PodCIDRs))
+	if pc := node.Spec.PodCIDR; pc != "" {
```
Comment here would be useful. For the uninitiated such as me this is a bit confusing :)

ldelossa replied:
add comment, lmk if this helps @joamaki I essentially link to the struct field comments which explain why those data access expressions look a bit magical.

joamaki commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -725,10 +725,12 @@ data:
   {{ fail "BGP was enabled, but no announcements were enabled. Please enable one or more announcements." }}
 {{- else if and .Values.bgp.enabled .Values.bgp.announce.loadbalancerIP }}
   bgp-announce-lb-ip: {{ .Values.bgp.announce.loadbalancerIP | quote }}
+{{- else if and .Values.bgp.enabled .Values.bgp.announce.podCIDR }}
```
Is it correct that one can only enable bgp-announce-lb-ip or bgp-announce-pod-cidr, but not both ("else if")?

christarazi replied:
Both can be enabled, we should change the template to allow that

ldelossa replied:
I think for simplicity I want to do this: 

```
{{- if and .Values.bgp.enabled (and (not .Values.bgp.announce.loadbalancerIP) (not .Values.bgp.announce.podCIDR)) }}
  {{ fail "BGP was enabled, but no announcements were enabled. Please enable one or more announcements." }}
{{- end }}

{{- if and .Values.bgp.enabled .Values.bgp.announce.loadbalancerIP }}
  bgp-announce-lb-ip: {{ .Values.bgp.announce.loadbalancerIP | quote }}
{{- end }}

{{-if and .Values.bgp.enabled .Values.bgp.announce.podCIDR }}
  bgp-announce-pod-cidr: {{ .Values.bgp.announce.podCIDR | quote }}
{{- end}}
```

which takes the form of: 
  guard_feature_configuration
  set_feature_parameter
  set_feature_parameter

where the guard ensures the feature can be enabled without issue.

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -49,18 +69,58 @@ func (s *Speaker) run() {
+			failed = true
+			ret = r
+		}
+		log.WithField("CIDRs", k.podCIDRs).Error("Failed to announce load balancer IP.")
```
Do the podCIDRs need to be printed in the log msg here?

ldelossa replied:
whoops, you're right.

ldelossa commented on Documentation/gettingstarted/bgp.rst:
```diff
@@ -14,10 +14,10 @@ BGP provides a way to advertise routes using traditional networking protocols
-<https://metallb.universe.tf/>`_ simple and effective implementation of IP
-allocation and the minimal BGP protocol support to do this. The configuration
-for Cilium is the same as MetalLB's configuration.
+``LoadBalancer`` IPs of ``Services`` and a kubernetes node's Pod CIDR range via BGP. 
```
@christarazi - updated these docs to mention the pod cidr announcements.

christarazi replied:
Awesome, thanks. 

Minor nit to keep consistency
```suggestion
``LoadBalancer`` IPs of ``Services`` and a Kubernetes node's Pod CIDR range via BGP. 
```

aanm commented on pkg/k8s/watchers/subscriber/service.go:
```diff
@@ -29,30 +29,42 @@ func (l *ServiceList) Register(s ServiceHandler) {
+func (l *ServiceList) NotifyAdd(svc *slim_corev1.Service) []error {
 	l.RLock()
 	defer l.RUnlock()
+	errs := make([]error, 0, len(l.subs))
```
nit: we are unnecessarily allocating memory in the best case scenario, this could be `var errs []error` to avoid that allocation. In the worst case scenario, which should not happen that often, it's fine to pay that cost by letting golang to allocate memory as it needs.

joamaki replied:
Similar thing also in subscriber/node.go in case you'll want to refactor these.

ldelossa replied:
Nice, I like that reasoning. Will update.

ldelossa replied:
Fixed all these including ones in node.go

aanm commented on pkg/k8s/watchers/subscriber/service.go:
```diff
@@ -29,30 +29,42 @@ func (l *ServiceList) Register(s ServiceHandler) {
+func (l *ServiceList) NotifyUpdate(oldSvc, newSvc *slim_corev1.Service) []error {
 	l.RLock()
 	defer l.RUnlock()
+	errs := make([]error, 0, len(l.subs))
```
same here

aanm commented on pkg/k8s/watchers/subscriber/service.go:
```diff
@@ -29,30 +29,42 @@ func (l *ServiceList) Register(s ServiceHandler) {
+func (l *ServiceList) NotifyDelete(svc *slim_corev1.Service) []error {
 	l.RLock()
 	defer l.RUnlock()
+	errs := make([]error, 0, len(l.subs))
```
same here

aanm commented on pkg/k8s/watchers/subscriber/service.go:
```diff
@@ -29,30 +29,42 @@ func (l *ServiceList) Register(s ServiceHandler) {
 
 // NotifyAdd notifies all the subscribers of an add event to a service.
-func (l *ServiceList) NotifyAdd(svc *slim_corev1.Service) {
+func (l *ServiceList) NotifyAdd(svc *slim_corev1.Service) []error {
```
I don't understand this refactoring. The caller of this function doesn't do anything with the returned errors.

christarazi replied:
I think we'll want to have this so that https://github.com/cilium/cilium/issues/15471 can be more robust and we can respond to errors if we need to. In this case, there may not be an immediate use case to using the returned errors, but it gives us flexibility in the future.

@ldelossa to confirm.

ldelossa replied:
@christarazi I do think this is the sentiment, to future proof this usage. Say if "ServiceList" was used else where the calling code could react to specific service event handler errors. 

@aanm what are your thoughts on this?

aanm replied:
Discussed offline

aanm commented on pkg/bgp/speaker/pod_cidr.go:
```diff
@@ -0,0 +1,75 @@
+)
+
+func (s *MetalLBSpeaker) withDraw() error {
+	var wg sync.WaitGroup // waitgroup here since we don't care about errors
```
If we don't care about errors we why do we have the ability to return an error? :-)

ldelossa replied:
yup, going to remove this from it's signature.

aanm commented on pkg/bgp/speaker/events.go:
```diff
@@ -17,18 +20,35 @@ type svcEvent struct {
+type nodeEvent struct {
+	// The following fields must be a pointers because they are not hashable
+	// (read comparable) in Go.
+	labels   *map[string]string
```
Can we have a comment of what `labels` are?

aanm commented on pkg/bgp/speaker/events.go:
```diff
@@ -49,18 +69,58 @@ func (s *Speaker) run() {
+			failed = true
+			ret = r
+		}
+		log.Error("Failed to announce load balancer IP.")
```
This message is not really helpful to the user as it doesn't answer the "Why?"

joamaki replied:
Aren't we logging this even when it succeeds?

ldelossa replied:
ack both comments, will fix.

aanm reviewed (approved):
LGTM overall, just some small nits that doesn't require a 2nd review from my side.

joamaki commented on pkg/bgp/speaker/pod_cidr.go:
```diff
@@ -0,0 +1,75 @@
+			eg.Go(func() error {
+				err := s.announce(sess, cidrs)
+				if err == nil {
+					log.WithField(logfields.CIDR, cidrs).Debug("Announced Pod CIDRs")
```
Would it be useful to log info about the session?

ldelossa replied:
Session is from the metallb api and has no informative data:

```
// Session gives access to the BGP session.
type Session interface {
	io.Closer
	Set(advs ...*bgp.Advertisement) error
}
```

ldelossa replied:
Actually kinda surprised to find that the session doesn't carry much useful info with it.

ldelossa replied:
@joamaki going to resolve this for now, but if I'm missing something feel free to un-resolve and ping me.

joamaki reviewed (approved):
Few nits, but other than that LGTM

qmonnet commented on Documentation/gettingstarted/bgp.rst:
```diff
@@ -60,7 +60,9 @@ Here are the required Helm values:
 .. parsed-literal::
 
-   helm install cilium |CHART_RELEASE| --set bgp.enabled=true --set bgp.announce.loadbalancerIP=true
+   helm install cilium |CHART_RELEASE| --set bgp.enabled=true --set bgp.announce.loadbalancerIP=true --set bgp.announce.podCIDR=true
```
Note: You may get a conflict from https://github.com/cilium/cilium/pull/17496 on this line.

ldelossa replied:
you called it @qmonnet  ;) I'll rebase.

ldelossa committed:
bgp_manager,speaker: avoid duplicate event processing

Prior to this commit it was possible for the BGP advertisement business
logic to replay stale events.

This scenario would happen when concurrently enqueuing new events while
re-queuing events which failed to be processed.

The re-queued event has the ability to get "replayed" after a
concurrently enqueued new event, leading to an ABA problem.

This commit adds a new type called "fence.Fencer" which keeps track of
the largest revision number seen for a particular UUID.

Both revision numbers and UUID's are Kubernetes (Etcd) values crafted
for this specific purpose.

Signed-off-by: Louis DeLosSantos <louis.delos@isovalent.com>

aanm commented on pkg/bgp/manager/subscriber.go:
```diff
@@ -17,73 +20,96 @@ import (
+		return err
+	}
+	return nil
+}
```
Wouldn't be "cleaner" to have a constructor that returns a `svcEvent` from a `*slim_corev1.Service`? :thinking:

ldelossa replied:
I personally like the pattern:

```
meta := Meta{}
meta.FromSlimService(...)
```

since the Go syntax allow is doing the allocation. I think its cleaner from the callers perspective.

It is also helpful since we do a lot of data conversion from Slim/native k8s, so tacking on a new "FromK8sService" is easy.

aanm replied:
It looks odd that if there's an error returned by lines 46, the object will be half-backed modified from the line 45.

ldelossa replied:
Okay that is fair, I'll leave this pattern to pure data objects, where side effects/errors cannot happen.

ldelossa replied:
Okay I moved this method to a "svcEventFromSlimService" constructor.

aanm commented on pkg/bgp/manager/subscriber.go:
```diff
@@ -160,10 +201,21 @@ func (m *Manager) reconcile(name string, svc *slim_corev1.Service) types.SyncSta
-	for _, k := range m.indexer.ListKeys() {
-		m.queue.Add(svcEvent(k))
+func (m *Manager) forceResync() error {
+	for _, s := range m.indexer.List() {
```
The `m.indexer.List()` will return what I believe a slice with a copy of all objects from this indexer. Thus it's probably cheaper to use `ListKeys()` instead.

ldelossa replied:
I'm under the impression the only objects in this indexer are service objects, is that true?

If that is the case, we do want all the service objects so we can convert them into svcEvent structures and requeue the events.

aanm replied:
The underlying issue I was trying to point out is the fact that it's returning a copy of all services, unnecessarily.

ldelossa replied:
But why unnecessarily ? We use the service objects returned to create the service event: 

https://github.com/cilium/cilium/pull/16525/commits/83df1288f2d23e27c06f18e48b3cda1c6f439cfc#diff-e24e587b0b67c16066b4f0d214a9e5fb3b457f763f02279cb15d65cb9c866fc2R212

ldelossa replied:
This becomes a moot point after backing out the fence changes from the Manager. Resolving.

aanm commented on pkg/bgp/manager/subscriber.go:
```diff
@@ -97,40 +162,60 @@ func (m *Manager) run() {
+				WithField("revision", e.meta.Rev).
+				Debug("Encountered stale event, will not process")
+			return types.SyncStateSuccess
+		}
```
This use case seems odd to me, are we expecting that many stale events that would require us from keeping a map with the Revision of a particular service? What's the harm if we re-process the event?

ldelossa replied:
I think you're right for this case, I'm now noticing that the Manager is asking the indexer for the latest resource state, which means an out of order event introduces not much harm.

this is however, not the case for the Speaker code. Currently the speaker code does not have an indexer and does not check against the latest updated state.

Therefore, I believe its possible for the speaker to advertise stale data by means of requeing an older message event behind a newer one. 

I could remove the fence from the Manager it appears.

christarazi replied:
+1, I don't think we need to use the fencer here because we can rely on client-go caches giving us the latest object.

aanm reviewed (changes requested):
I only reviewed the last commit.

christarazi commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,84 @@
+package fence
```
Copyright header

christarazi commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,84 @@
+// a slim k8s ObjectMeta and stores it at the memory
+// pointed to by m.
+func (m *Meta) FromSlimObjectMeta(om *slim_metav1.ObjectMeta) {
+	rev, err := strconv.ParseUint(om.ResourceVersion, 10, 64)
```
FWIW, Kubernetes states that we should not do this: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions

ldelossa replied:
This will be resolved when we do not queue the event details and instead retrieve data from the "client-go" caches.

christarazi commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,84 @@
+		// panic justification: our stale event avoidance (correctness) logic depends
+		// on this, if ObjectMeta changes under us, crash immediately
+		// and fix this logic.
+		panic("ObjectMeta.ResourceVersion must be parsible to Uint64")
```
If this happens, we don't want to crash Cilium. Can we return the error instead?

ldelossa replied:
@christarazi coupled with your comment here: https://github.com/cilium/cilium/pull/16525#discussion_r731381971

I would actually feel better if we did crash the process. 

The reason being, right now we assume (pretty reliably) that this field is coming from etcd. This is well documented and is even in several books about k8s programming. 

However, k8s devs want to maintain the ability to swap out that implementation (which IMO is not very likely without some advanced notice), thus they advertise that these values may not exist over time or be a different type. 

I'd like to panic here so that when a Cilium version is being tested against a k8s upgrade path, if the 'ResourceVersion' field changes, the tester doesn't loose this information in the logs. 

Instead, I'd feel better if the process was to crash and they inform us of this and we immediately become aware that the correctness of our algo is compromised.

christarazi replied:
I understand that it gives us immediate feedback if the `ResourceVersion` field changes. However, there's a balance to be found here. On one hand, we get immediate feedback when this changes. The cost of this immediate feedback is Cilium being stuck in a constant crashloopbackoff because each it comes up, it loads the K8s caches, it gets the "bad" object, and causes a panic. Restarts, loads cache, panic, and repeat. On the other hand, we fail to reconcile this event  and instead we warn loudly (error level). Users can alert on errors present in the logs so there's some relatively quick feedback with that as well, and we don't completely halt the user's environment. 

I'm not sure I'm comfortable with that. We have precedence of printing an error log when we don't expect the right object type for example (grep the code for `%T`). However, I'll raise this for discussion on #development.

ldelossa replied:
okay @christarazi I backed this out and swapped it with a log of the error.

christarazi commented on pkg/bgp/speaker/speaker.go:
```diff
@@ -66,13 +68,18 @@ type Opts struct {
 type MetalLBSpeaker struct {
+	// Our speaker requeues our own event structures on failure.
+	// Use a fence to avoid replaying stale events.
+	fence.Fencer
```
Nit: I prefer not embedding it, and rather making it an explicit field. I don't see a strong reason to embed this.

ldelossa replied:
Want to avoid the api. 

"s.fencer.fence" and promote the method to "s.fence".

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -7,16 +7,18 @@ package speaker
 // svcEvent holds the extracted fields from a K8s service event
 // which are of interest to the BGP package.
 type svcEvent struct {
+	fence.Meta
```
Same here re: embedding

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -106,12 +111,33 @@ func (s *MetalLBSpeaker) do(key interface{}) types.SyncState {
+		if s.Fence(k.Meta) {
+			l.WithField("uuid", k.Meta.UUID).
+				WithField("type", "service").
+				WithField("revision", k.Meta.Rev).
```
```
l.WithFields(logrus.Fields{
  "uuid": ...,
  "type": ...,
})...

christarazi commented on pkg/bgp/speaker/events.go:
```diff
@@ -84,13 +87,15 @@ func (s *MetalLBSpeaker) run(ctx context.Context) {
 		}
+		// if queue.Add(key) is called previous to this invocation the event
+		// is requeued, else it is discarded from the queue.
+		s.queue.Done(key)
```
Why is this diff necessary? I can see value in the additional comment, but not moving the line

ldelossa replied:
Shouldn't we call "Done" in the event of case types.SyncStateSuccess, types.SyncStateReprocessAll: and case types.SyncStateError:

not just in the error state?

christarazi replied:
Then we need to remove the duplicated `defer Done()` inside `do()`.

ldelossa replied:
Yup that's correct, thanks I did this in the Manager's code, but missed it in Speaker.

ldelossa replied:
fixed.

christarazi reviewed (commented):
I reviewed the last commit. As per our offline discussion, we were talking about the viability of the map approach storing all object instance's UUIDs to their `ResourceVersion`.

The summary of our conversation is that we need to do this in order to tackle the out-of-order problem because we can't utilize the client-go caches which can retrieve us the latest state of an object. The reason we can't utilize that is because we already store the services in the ServiceCache internally and having the client-go cache (indexer) would be duplicating memory.

The primary concern I had with the map approach is that it doesn't have a corresponding `delete()` so it's ever-increasing. One idea to is to implement the delete handler for each event type and upon successful delete handling, remove it from the fencer. Once we handle the ever-increasing map problem, then this approach is fine by me. 

It's worth noting that this still keeps mappings for all active instances, which isn't too bad considering that if we used the client-go approach, we'd be storing full objects anyway. Until we implement a better approach to K8s event handling like utilizing the client-go caches exclusively, this seems like a good enough stop-gap.

joamaki commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,84 @@
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Fencer provides a method set to prevent
```
perhaps run `fmt -w80 --prefix=//` on the comments here as they're a bit randomly line-wrapped?

ldelossa replied:
there's a new line char, I just cut it off short.

ldelossa replied:
fixed.

joamaki commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,84 @@
+//
+// Fencer will keep track of the last seen revision (monotonically increasing event id)
+// for each seen UUID (globally unique identifier for a resource producing an event.)
+type Fencer map[string]uint64
```
Are we sure this isn't concurrently accessed? I can see `lock.Mutex` at least in MetalLBSpeaker, which might be protecting this, but hard to say.

How big can this grow? Do we need to prune it?

ldelossa replied:
will double check but fairly certain we only access this array in our synchronous event processing loop. Events concurrently go into a queue, but only a single thread is consuming and processing. The single thread that consumes is where this map is accessed. 

@christarazi brought up concerns about pruning. going to work that into this commit.

ldelossa replied:
pruning implemented. Also confirmed that map is not accessed concurrently.

joamaki commented on pkg/bgp/manager/subscriber.go:
```diff
@@ -130,22 +162,31 @@ func (m *Manager) run() {
-	defer m.queue.Done(event)
+	l := log.WithFields(
+		logrus.Fields{
+			"component": "Manager.do",
```
s/do/process/  ?

christarazi commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2016-2021 Authors of Cilium
```
Nit: this file only existed as of a few days ago :wink:

ldelossa replied:
fixed

christarazi commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,92 @@
+	"strconv"
+
+	slim_metav1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Nit: import ordering, newlines

ldelossa replied:
fixed

christarazi commented on pkg/bgp/fence/fence.go:
```diff
@@ -0,0 +1,92 @@
+//
+// Fencer will keep track of the last seen revision (monotonically increasing event id)
+// for each seen UUID (globally unique identifier for a resource producing an event.)
+type Fencer map[string]uint64
```
I'm a bit surprised that we don't need to do a `make()` on the map. Do you know why that is? Or is there a nil reference waiting to happen?

ldelossa replied:
This is a type definition, not a global variable definition.

christarazi replied:
But when we first use `Fencer`, doesn't the map need to be initialized?

ldelossa replied:
Fencer *is* the map. So it's initialized like any other map:

```
fence := Fencer{}
``` 
or
```
var fence fence.Fencer = make(map[string]uint64, 0, 0) 
```
or
```
fence := make(fence.Fencer, 0, 0)
```

all of them are valid, the first one is what occurs in this commit. 

The definitions are in the constructors which utilize a fence, its not a global fence declared in this package.

christarazi replied:
Ah I see, I forgot that `fence := Fencer{}` also initializes a map similar to `make()`. Thanks

ldelossa replied:
yup, that's a hard one to remember since it's a thing the Go compiler just decided was a good idea to do.

christarazi reviewed (commented):
A few more small comments. It's worth going through the code and ensuring all the log lines are consistently formatted (capitalization, etc).
